Support normalization of $SRC_DIR

Rocket's CI has been broken for a while because of UI test diffs like this:

```
 error[E0277]: the trait bound `usize: rocket::response::Responder<'_>` is not satisfied
-  --> $DIR/responder-types.rs:38:13
-   |
-38 | fn foo() -> usize { 0 }
-   |             ^^^^^ the trait `rocket::response::Responder<'_>` is not implemented for `usize`
-   |
-   = note: required by `rocket::handler::<impl rocket::Outcome<rocket::Response<'r>, rocket::http::Status, rocket::Data>>::from`
+   --> $DIR/responder-types.rs:38:13
+    |
+38  | fn foo() -> usize { 0 }
+    |             ^^^^^ the trait `rocket::response::Responder<'_>` is not implemented for `usize`
+    | 
+   ::: /Users/runner/runners/2.159.2/work/1/s/core/lib/src/handler.rs:202:20
+    |
+202 |     pub fn from<T: Responder<'r>>(req: &Request<'_>, responder: T) -> Outcome<'r> {
+    |                    ------------- required by this bound in `rocket::handler::<impl rocket::Outcome<rocket::Response<'r>, rocket::http::Status, rocket::Data>>::from`
```

The new annotations appear to be introduced by rust-lang/rust#64151, in which `$SRC_DIR` has been used to normalize similarly added paths in the output (e.g. https://github.com/rust-lang/rust/pull/64151/files#diff-4b63aaabf76aa70969c04a1a4185c421R7).

`$SRC_DIR` normalization was added to `rustc` in rust-lang/rust#50943 but it doesn't appear to be supported in compiletest - this means that it is not currently possible for these errors to be properly tested both in CI and locally, as each run ends up with files in a different path.

If the same change from rust-lang/rust#50943 is likely to be usable (and accepted) here, I'd be happy to make that pull request.

Yeah, we'd accept that PR

On Tue, Nov 5, 2019, 8:15 PM Jeb Rosen <notifications@github.com> wrote:

> Rocket's CI has been broken for a while because of UI test diffs like this:
>
>  error[E0277]: the trait bound `usize: rocket::response::Responder<'_>` is not satisfied
> -  --> $DIR/responder-types.rs:38:13
> -   |
> -38 | fn foo() -> usize { 0 }
> -   |             ^^^^^ the trait `rocket::response::Responder<'_>` is not implemented for `usize`
> -   |
> -   = note: required by `rocket::handler::<impl rocket::Outcome<rocket::Response<'r>, rocket::http::Status, rocket::Data>>::from`
> +   --> $DIR/responder-types.rs:38:13
> +    |
> +38  | fn foo() -> usize { 0 }
> +    |             ^^^^^ the trait `rocket::response::Responder<'_>` is not implemented for `usize`
> +    |
> +   ::: /Users/runner/runners/2.159.2/work/1/s/core/lib/src/handler.rs:202:20
> +    |
> +202 |     pub fn from<T: Responder<'r>>(req: &Request<'_>, responder: T) -> Outcome<'r> {
> +    |                    ------------- required by this bound in `rocket::handler::<impl rocket::Outcome<rocket::Response<'r>, rocket::http::Status, rocket::Data>>::from`
>
> The new annotations appear to be introduced by rust-lang/rust#64151
> <https://github.com/rust-lang/rust/pull/64151>, in which $SRC_DIR has
> been used to normalize similarly added paths in the output (e.g.
> https://github.com/rust-lang/rust/pull/64151/files#diff-4b63aaabf76aa70969c04a1a4185c421R7
> ).
>
> $SRC_DIR normalization was added to rustc in rust-lang/rust#50943
> <https://github.com/rust-lang/rust/pull/50943> but it doesn't appear to
> be supported in compiletest - this means that it is not currently possible
> for these errors to be properly tested both in CI and locally, as each run
> ends up with files in a different path.
>
> If the same change from rust-lang/rust#50943
> <https://github.com/rust-lang/rust/pull/50943> is likely to be usable
> (and accepted) here, I'd be happy to make that pull request.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/laumann/compiletest-rs/issues/198?email_source=notifications&email_token=AAMK6SDZD2L3T2TCH3H6METQSJAEVA5CNFSM4JJPKBE2YY3PNVWWK3TUL52HS4DFUVEXG43VMWVGG33NNVSW45C7NFSM4HXDPWTA>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAMK6SH2Y4LYORYQYZOF3T3QSJAEVANCNFSM4JJPKBEQ>
> .
>

--- a/src/runtest.rs
+++ b/src/runtest.rs
@@ -2616,6 +2616,10 @@ actual:\n\
         let parent_dir = self.testpaths.file.parent().unwrap();
         normalize_path(parent_dir, "$DIR");
 
+        let src_dir_str = std::env::current_dir().unwrap();
+        let src_dir = Path::new(&src_dir_str);
+        normalize_path(src_dir, "$SRC_DIR");
+
         // Paths into the build directory
         let test_build_dir = &self.config.build_base;
         normalize_path(test_build_dir, "$TEST_BUILD_DIR");

--- a/src/runtest.rs
+++ b/src/runtest.rs
@@ -2616,9 +2616,10 @@ actual:\n\
         let parent_dir = self.testpaths.file.parent().unwrap();
         normalize_path(parent_dir, "$DIR");
 
-        let src_dir_str = std::env::current_dir().unwrap();
-        let src_dir = Path::new(&src_dir_str);
-        normalize_path(src_dir, "$SRC_DIR");
+        if let Ok(src_dir_str) = std::env::var("CARGO_MANIFEST_DIR") {
+            let src_dir = Path::new(&src_dir_str);
+            normalize_path(src_dir, "$SRC_DIR");
+        }
 
         // Paths into the build directory
         let test_build_dir = &self.config.build_base;

